### PR TITLE
Biosample to metadata master

### PR DESCRIPTION
**Pull request by mbaudis**
This is a skeleton implementation of the ```Biosample``` object, to provide a hook for e.g. reads/variants objects.

### CODE REVIEW

**Comment by lh3** [comment]
Is there any uniqueness requirement on `Biosample.name`?

**Comment by mbaudis** [comment]
@lh3 Wouldn't think so. While the schema can't express this (?), this goes into the general "id definition problem". My take on this (expressed partially in various discussions) is:

* id: required; should be sane & at least locally unique identifier, usually something "readable" (GSM2492834, BRCA-2015-DCIS0012 ...)
* guid: uuidv4, computer generated at creation 9recommended)
* name: a not necessarily "clean", "descriptive id"
* description: whatever free, unstructured annotation the user wants to use; explicitly not containing attribute::value pairs

The distinction between "name" and "id" seems a bit arbitrary/duplicative, it reflects common usage (e.g. an id + a sample name), while the guid is the enhancement.

File: src/main/resources/avro/metadata.avdl [hunk 1]
```diff
@@ -92,6 +92,60 @@ record Experiment {
...
+
+  /**
+  External identifiers for sample, e.g. EBI BioSamples accession.
+  */
+  array<string> externalIds;
+
+  /**
+  A description of the biosample.
+  */
+  union { null, string } description = null;
```
Comment (jeromekelleher) [inline src/main/resources/avro/metadata.avdl]: Looks like a duplicated ``description`` field @mbaudis.

**Comment by mbaudis** [comment]
@jeromekelleher Oops; fixed. Late here ...

**Comment by jeromekelleher** [comment]
Looks sensible to me @mbaudis. I'd make the same comments I did on #399 about IDs vs names. Having IDs be server generated and opaque is very useful implementation-wise, and we'd certainly struggle to replace that in the reference server. I think ``name`` is a generally useful thing for users. I want somewhere to put 'NA12878' or '1kg-phase3'. 

I  agree about GUIDs, but I think this is probably the sort of thing we should implement after we sync up the schemas and start the new development model. Implementation will make it a lot easier to see what works and what doesn't.

**Comment by lh3** [comment]
I asked this because the BioSample database appears to require global naming uniqueness (based on the existence of many related but different NA12878 "sample name": PB_NA12878-sc-1646934, E-MTAB-1884:NA12878, NA12878_30x, Platinum3, NA12878-Human, NA12878-sc-2011-01-04-7605, ... which I found messy). It is good that you are not thinking of the same type of "name".

**Comment by helenp** [comment]
@lh3 the samples you flag there are from a source database that we import to BioSamples that does require naming uniqueness - they are are from ArrayExpress - Biosamples itself does not, just needs unqiue ids. As BioSamples is an aggregator of some existing resources as well as getting direct submissions there are some data source specific artifacts. ArrayExpress will switch to using BioSamples ids to get around this problem in future. I agree with your desire to have a name as well as id.

**Comment by mbaudis** [comment]
I think there is some consent about having an id (which is controlled system wide at the resource provider) and additionally a name. The id may be computer generated or curated, but itself would be required & suitable for referencing, at least in a local context.

On the other hand, objects should have a guid, created at object generation time, in a consistent, "anonymous", collision free format. My vote right now would be for UUIDv4. Though I would vote for this to be a required feature, I see at least 2 points which would have to be discussed:
* privacy implications, if the guid is transmitted within otherwise anonymous objects but may lead to de-anonymization when co-occurring in another place
* overhead; i.e. do you really want this to be attached to every reads/variants ... object

Another concept would be to make the ```id``` attributes being the guid itself, and the name then being the "human readable/local specific" id. Other legacy ids could then be dumped not an ```aliases``` | ```accessions``` list.

So this is something for experts like @helenp to opine on ...

**Comment by jeromekelleher** [comment]
@mbaudis --- we should also consider the possibility of GUIDs being cryptographic content hashes which are attached to every object (reads, readgroups, variants, variant sets, datasets, etc). This is a lot more useful than just locally creating UUIDs in my opinion as it enables cross-site comparisons of data, real guarantees of data identity, versioning, and many other benefits. There has been a _lot_ of discussion about this --- see  #167.

**Comment by mbaudis** [comment]
@jeromekelleher Yes, but this would be a content/version id. For tracking of ```wet``` objects, a static GUID is much more practical, since changing its information content wouldn't break the id. Has been part of metadata discussions before; and I've mentioned the additional versioning/hash option/requirement in previous comments. We also have ```recordUpdateTime``` for sorting objects with the same guid.

Also, I'm more for offering a larger selection of well-defined attributes with clear roles, instead of "users" (i.e. local implementation developers) designing their own workarounds.

**Comment by mbaudis** [comment]
... ```recordUpdateTime``` being implemented in the metadata branch.

```
    The time at which this record was created.
    Format: ISO 8601, YYYY-MM-DDTHH:MM:SS.SSS (e.g. 2015-02-10T00:03:42.123Z)
    TODO: The format definition is temporary and will be updated to
    the upcoming AVRO time format (timestamp-millis) when this is supported.
    */
    string recordCreateTime;

    /**
    The time at which this record was last updated.
    Format: ISO 8601, YYYY-MM-DDTHH:MM:SS.SSS (e.g. 2015-02-10T00:03:42.123Z)
    */
    string recordUpdateTime;
```

**Comment by jeromekelleher** [comment]
@mbaudis --- fair enough, I hadn't thought about wet objects. Perhaps these could have uuids and a content hash? It's still useful to compare them for content equality quickly I think.

**Comment by mbaudis** [comment]
@jeromekelleher Yes, exactly my opinion (id + name + guid + hash).

**Comment by jeromekelleher** [comment]
@mbaudis Ah, OK. Sorry I'm a bit slow on catching up on all the metadata work!

**Comment by mbaudis** [comment]
@jeromekelleher @diekhans @dglazer @helenp Following the discussions about id | guid | name | accessions, I agree that we should set up the general id attribute procedure first before specifying something beyond ```id``` + ```name``` in master.
I've changed the Biosample PR to just include ```id```, ```name```, ```description```. Pls. vote, for moving this into master.

**Comment by jeromekelleher** [comment]
I think this sort of change needs to be done under the new development model we've agreed on --- we need to implement Biosample in the reference server and compliance suite before it can be added to master in the schemas. Changes to master going on right now are just bug-fixes to get us to the point where we can sync up the three repositories, get the current schema into a more consistent state and tag a version we can work from. 

So, I'm -1 on adding a change like this to master right now. BioSample sounds like an ideal topic branch though...

**Comment by diekhans** [comment]
@mbaudis, what is the motivation of doing this PR now?  I know the background
clarification related the Heng's proposed changes to variants.  If  there
isn't a more immediate need, then perhaps it should go into the topic branch
that implements that change.

**Comment by mbaudis** [comment]
While driven by the Variants discussion, ```Biosample``` is a general object on a level outside of ref/var. It should not be implemented there.

The PR is, as the integration of ```Dataset``` has been, part of a "bottom up" piecemeal integration of ongoing branch developments. A PR against variants would lead to possible collisions with other branches (but anyway, just wouldn't be *clean*).

**Comment by mbaudis** [comment]
I have to reiterate my push for getting ```Biosample``` into the master branch. It represents the core reference object for any analysis object.

**Comment by diekhans** [comment]
We need to get it implemented in the reference server and
compliance suites.  This should probably go in as part of a
work that actually needs it.

**Comment by mbaudis** [comment]
@diekhans So who will take this up?

**Comment by mbaudis** [comment]
I've changed from ```Biosample``` to ```BioSample```, as this seems to be the preferred spelling of the record's name (e.g. https://github.com/ga4gh/schemas/pull/427); also goes with ```AnalysisSample``` (https://github.com/ga4gh/schemas/pull/391).

**Comment by mbaudis** [comment]
I've added fixes in ```reads.avdl``` and ```variants.avdl``` to get the record names in sync (```bioSampleId``` instead of ```sampleId```).

**Comment by mbaudis** [comment]
Closing this since work has been moved to https://github.com/mbaudis/schemas =>

**Pull request closed without merging**